File: solutions/ocaml/triangle/1/triangle.ml
let is_equilateral a b c =
    a > 0 && a = b && b = c

let is_isosceles a b c =
    (a = b && a + b > c || a = c && a + c > b || b = c && b + c > a) && a > 0 && b > 0 && c > 0

let is_scalene a b c =
    a > 0 && b > 0 && c > 0 && not (is_isosceles a b c) && a + b > c && a + c > b && b + c > a

File: solutions/kotlin/triangle/1/src/main/kotlin/Triangle.kt
class Triangle<out T : Number>(val a: T, val b: T, val c: T) {

    // TODO: Implement proper constructor
    
    init {
        val x = a.toDouble()
        val y = b.toDouble()
        val z = c.toDouble()
        require(x + y > z && x + z > y && y + z > x)
    }

    val isEquilateral: Boolean
    get() {
        return a.toDouble() > 0 && a == b && b == c
    }
    val isIsosceles: Boolean
    get() {
        return a == b || a == c || b == c
    }
    val isScalene: Boolean = !isIsosceles
}
